lib/changelog.js

```diff
@@ -0,0 +1,237 @@
+var Async = require('async');
+var Config = require('getconfig');
+var Utils = require('./utils');
+var Querystring = require('querystring');
+
+var internals = {};
+
+internals.dependencyRegex = /(?:update|upgrade) (\w+)\/(\w+) to (.*) from (.*)/i;
+
+internals.hapiOrg = 'hapijs';
+
+internals.perPage = 100;
+
+internals.requestOptions = {
+    headers: {
+        'user-agent': 'hapijs.com',
+        'authorization': 'token ' + Config.githubToken
+    },
+    json: true
+};
+
+exports.methods = [];
+
+exports.methods.push({
+    name: 'changelog.milestones',
+    method: function (callback) {
+
+        var params = Querystring.encode({
+            state: 'closed',
+            per_page: internals.perPage
+        });
+
+        return Utils.downloadAllPages('https://api.github.com/repos/hapijs/hapi/milestones?' + params, internals.perPage, internals.requestOptions, function (err, milestones) {
+
+            if (err) {
+                return callback(err);
+            }
+
+            var compare = function (a, b) {
+
+                if (a.id < b.id) {
+                    return 1;
+                }
+                if (a.id > b.id) {
+                    return -1;
+                }
+                return 0;
+            };
+
+            callback(null, milestones.sort(compare));
+        });
+    },
+    options: {
+        cache: {
+            expiresIn: Utils.oneHour,
+            staleIn: Utils.fifteenMinutes,
+            staleTimeout: 1000
+        },
+        generateKey: function () {
+
+            return 'changelogMilestones';
+        }
+    }
+});
+
+exports.methods.push({
+    name: 'changelog.build',
+    method: function (request, callback) {
+
+        var methods = request.server.methods.changelog;
+        var changelog = { };
+
+        var getDependencyIssues = function (issue, cb) {
+
+            methods.getDependencyIssues(request, issue.title, function (result, err) {
+
+                if (err) {
+                    return cb(err);
+                }
+
+                if (!result) {
+                    return cb(null);
+                }
+
+                issue.pluginIssues = {
+                    repo: internals.hapiOrg + '/' + result.repo,
+                    issues: result.issues
+                };
+
+                cb(null);
+            });
+        };
+
+        var checkForDependencies = function (issues, cb) {
+
+            if (issues.length > 0) {
+
+                return Async.eachLimit(issues, 10, getDependencyIssues, function (err) {
+
+                    if (err) {
+                        return cb(err);
+                    }
+
+                    cb(null);
+                });
+            }
+
+            cb(null);
+        };
+
+        var getIssues = function (milestone, cb) {
```

Comment:
Instead of having this inside a giant function, you can put this into  `internals.checkForDependencies` and move it outside.